### PR TITLE
[RFC] Move part of syntax/vim.vim to syntax/vim/generated.vim

### PR DESCRIPTION
**Pull request by ZyX-I**
First two commits aim at providing necessary information in a lua-friendly format which does not require implementing parser. Commands are already in a friendly format, these commits move autocommands and options there as well.

Autocommand names move has additional benefit of avoiding duplicating data (enum and actual autocommands list; originally this commit was created because I needed list of autocommands for my parser).

Options were moved because I think that the result is more readable then the initial value; and it is also easier to convert from lua dictionary then from C structure (I constantly hear rumors about reworking option facility). Functions are taken directly from eval.c.

File is not actually generated or used because I don’t know which would be the correct variant for using it: creating `runtime/CMakeLists.txt` file or using the top-level one.

### CODE REVIEW

**Comment by ZyX-I** [comment]
Build failed due to unrelated reason:

```
W: Failed to fetch http://www.rabbitmq.com/debian/dists/testing/main/i18n/Translation-en Unable to connect to www.rabbitmq.com:http:

E: Some index files failed to download. They have been ignored, or old ones used instead.

The command "if [ $TRAVIS_OS_NAME = linux ]; then sudo apt-mark hold oracle-java7-installer oracle-java8-installer; sudo apt-get update; elif [ $TRAVIS_OS_NAME = osx ]; then brew update; fi" failed and exited with 100 during .

Your build has been stopped.
```

. Will restart it.

**Comment by ZyX-I** [comment]
There are also differences between generated keywords for commands and currently used keywords: I have not looked at all of them, but e.g. `loadk` is represented as `loadkeymap` and `loadk` in two adjacent lines (both variants are present in the middle) which is rather strange, my file uses `loadk[emap]` instead. Also note that some abbreviations do not exist officially (not mentined in `:h`), though they work and will be highlighted with the generated file.

**Comment by splinterofchaos** [comment]
This definitely looks like an improvement to me. :+1: 

I wonder if a `runtime/doc` file shouldn't be generated to list the undocumented (and documented) abbreviations, just linking them to their non-abbreviated `:hep` sections.

**Comment by ZyX-I** [comment]
This is not yet an improvement. I need to know where to put runtime file generation: either to `/runtime/CMakeLists.txt` or to `/CMakeLists.txt`.

**Comment by ZyX-I** [comment]
Wondering what this may mean:

```
$ git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim
Cloning into 'neovim/neovim'...
fatal: read error: Connection reset by peer
The command "eval git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim" failed. Retrying, 2 of 3.
Cloning into 'neovim/neovim'...
remote: Counting objects: 3365, done.
remote: Compressing objects: 100% (2600/2600), done.
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

```
$ git clone --depth=50 git://github.com/neovim/neovim.git neovim/neovim
Cloning into 'neovim/neovim'...
remote: Counting objects: 3365, done.
remote: Compressing objects: 100% (2600/2600), done.
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

(first is gcc failure, second is gcc-32 failure).

**Comment by ZyX-I** [comment]
Looks like build failed again due to DDoS or something like this.

**Comment by ZyX-I** [comment]
Since I got no replies about where I should place the runtime file generation it was placed to `runtime/CMakeLists.txt`. Help tags file generation was also moved there, and it is now part of the `make all` and not part of the installation step.

**Comment by ZyX-I** [comment]
I have rebased on the current master. Is anybody going to merge this? With constant changes to various option defaults rebasing is not very fun: got 236 lines in `git diff -U0 autovimsyn..autovimsyn-old src/nvim/options.lua`, and there is no conversion script (not very trivial to write).

**Comment by justinmk** [comment]
I missed this somehow, so thanks for the reminder. Considering we've had zero problems since @ZyX-I converted the ex_cmds list to a lua generator, I have no compunctions about merging this.

File: runtime/syntax/vim.vim [hunk 1]
```diff
@@ -64,24 +33,20 @@ syn match   vimTermOption contained	"t_&8"
...
 " Highlight commonly used Groupnames {{{2
 syn keyword vimGroup contained	Comment Constant String Character Number Boolean Float Identifier Function Statement Conditional Repeat Label Operator Keyword Exception PreProc Include Define Macro PreCondit Type StorageClass Structure Typedef Special SpecialChar Tag Delimiter SpecialComment Debug Underlined Ignore Error Todo 
 
 " Default highlighting groups {{{2
 syn keyword vimHLGroup contained	ColorColumn Cursor CursorColumn CursorIM CursorLine CursorLineNr DiffAdd DiffChange DiffDelete DiffText Directory ErrorMsg FoldColumn Folded IncSearch LineNr MatchParen Menu ModeMsg MoreMsg NonText Normal Pmenu PmenuSbar PmenuSel PmenuThumb Question Scrollbar Search SignColumn SpecialKey SpellBad SpellCap SpellLocal SpellRare StatusLine StatusLineNC TabLine TabLineFill TabLineSel Title Tooltip VertSplit Visual WarningMsg WildMenu
 syn match vimHLGroup contained	"Conceal"
+<<<<<<< HEAD
 syn keyword vimOnlyHLGroup contained	VisualNOS
 syn keyword nvimHLGroup contained	EndOfBuffer TermCursor TermCursorNC
+=======
```
Comment (justinmk) [inline runtime/syntax/vim.vim]: bad merge

**Comment by justinmk** [comment]
@ZyX-I There is a mismerge in the syntax file, but after that feel free to merge this.

**Comment by ZyX-I** [comment]
Fixed mismerge.

**Comment by ZyX-I** [comment]
Also note that for some reason I forgot to remove vimFuncName definitions from vim.vim. This was fixed alongside with the mismerge.

**Comment by fmoralesc** [comment]
I don't think this handles the `vimOnly*` groups, which would cause some reversion (imho) from #2902.

**Comment by ZyX-I** [comment]
@fmoralesc `vimOnly` groups will now be restored. The only thing which

**Comment by fmoralesc** [comment]
@ZyX-I Great. LGTM, then ;)

**Comment by Pyrohh** [comment]
I did an installation after this was merged, and noticed the directory structure has changed, was that intentional?
```
/usr/local/share/nvim $ tree -L 2
.
|-- doc
|   |-- makehtml.awk
|   `-- maketags.awk
|-- macros
|   `-- less.sh
`-- runtime
    |-- CMakeFiles
    |-- CMakeLists.txt
    |-- autoload
    |-- bugreport.vim
    |-- cmake_install.cmake
    |-- colors
    |-- compiler
    |-- delmenu.vim
    |-- doc
    |-- filetype.vim
    |-- ftoff.vim
    |-- ftplugin
    |-- ftplugin.vim
    |-- ftplugof.vim
    |-- indent
    |-- indent.vim
    |-- indoff.vim
    |-- keymap
    |-- macmap.vim
    |-- macros
    |-- makemenu.vim
    |-- menu.vim
    |-- mswin.vim
    |-- optwin.vim
    |-- plugin
    |-- print
    |-- rgb.txt
    |-- scripts.vim
    |-- synmenu.vim
    |-- syntax
    `-- vimrc_example.vim

15 directories, 22 files
/usr/local/share/nvim $
```

I'm not sure if all of these are issues:
- `less.sh` should be located in `$VIMRUNTIME/macros` according to `help :less`
- the following are duplicated from `runtime/`: `less.sh` and `make(html|tags).awk`
- CMakeFiles, CMakeLists.txt, and cmake_install.cmake probably shouldn't be installed

**Comment by ZyX-I** [comment]
@Pyrohh You are right, I have leaked some errors in the cmake script.

**Pull request merged**